### PR TITLE
Fix range header bounds check in SetRange()

### PR DESCRIPTION
**Pull request by donatello**
Should fix https://github.com/minio/minio/issues/4243

### CODE REVIEW

File: request-headers.go [hunk 1]
```diff
@@ -89,17 +89,18 @@ func (c RequestHeaders) SetRange(start, end int64) error {
 	case start > 0 && end == 0:
 		// Read everything starting from offset 'start'. `bytes=N-`
 		c.Set("Range", fmt.Sprintf("bytes=%d-", start))
-	case start > 0 && end > 0 && end >= start:
+	case start > 0 && end >= start:
 		// Read everything starting at 'start' till the 'end'. `bytes=N-M`
```
Comment (harshavardhana) [inline request-headers.go]: Won't this still retun an error if start is 0 and end > 0 ?

File: request-headers.go [hunk 2]
```diff
@@ -84,22 +84,21 @@ func (c RequestHeaders) SetModified(modTime time.Time) error {
...
-		// Read everything until the 'end'. `bytes=-N`
+		// Read last '-end' bytes. `bytes=-N`
 		c.Set("Range", fmt.Sprintf("bytes=%d", end))
 	case start > 0 && end == 0:
 		// Read everything starting from offset 'start'. `bytes=N-`
 		c.Set("Range", fmt.Sprintf("bytes=%d-", start))
-	case start > 0 && end > 0 && end >= start:
+	case start >= 0 && end >= start:
 		// Read everything starting at 'start' till the 'end'. `bytes=N-M`
 		c.Set("Range", fmt.Sprintf("bytes=%d-%d", start, end))
```
Comment (harshavardhana) [inline request-headers.go]: Can you test if 0-0 indeed reads the whole object?

Comment (donatello) [inline request-headers.go]: It does not.

**Comment by donatello** [comment]
There are two conflicting cases here:

1. Read 1 byte starting at the first byte - in the current implementation this is supported. The interpretation of `start==0 && end==0` is `bytes=0-0` (this is correct according to RFC). 
2. Read whole file starting from byte 0 - in the current implementation this is not supported. This is the same as not setting the header at all - so the `SetHeader` method should not be called.

**Comment by donatello** [comment]
If we implement in the opposite way, then:

`start == 0 && end == 0` would generate header as `bytes=0-` meaning read whole file from byte 0.

However, we lose the ability to express "read 1 byte starting from byte 0". So I chose to go with the latter. Still a better `SetHeader` API could be constructed.

File: request-headers.go [hunk 3]
```diff
@@ -83,23 +83,28 @@ func (c RequestHeaders) SetModified(modTime time.Time) error {
...
 		c.Set("Range", fmt.Sprintf("bytes=%d", end))
-	case start > 0 && end == 0:
-		// Read everything starting from offset 'start'. `bytes=N-`
+	case 0 < start && end == 0:
+		// Read everything starting from offset
+		// 'start'. `bytes=N-`.
 		c.Set("Range", fmt.Sprintf("bytes=%d-", start))
-	case start > 0 && end > 0 && end >= start:
-		// Read everything starting at 'start' till the 'end'. `bytes=N-M`
+	case 0 <= start && start <= end:
```
Comment (harshavardhana) [inline request-headers.go]: This looks good, can you add some tests for this as well? @donatello

**Comment by harshavardhana** [comment]
> If we implement in the opposite way, then:
> 
> start == 0 && end == 0 would generate header as bytes=0- meaning read whole file from byte 0.
> 
> However, we lose the ability to express "read 1 byte starting from byte 0". So I chose to go with the approach in the previous comment. Still a better SetHeader API could be constructed.

This is good serves the purpose here.

**Comment by donatello** [comment]
@harshavardhana Tests added.

File: request-headers_test.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
+package minio
```
Comment (harshavardhana) [inline request-headers_test.go]: Add license header..

Comment (donatello) [inline request-headers_test.go]: Done

**Pull request merged**